lib/node.js

```diff
@@ -533,11 +580,492 @@ class Node {
+   * 
+   * @param {Parameter} parameter - The new parameter.
+   * @return {rclnodejs.rcl_interfaces.msg.SetParameterResult} - The result of the operation.  
+   */
+  setParameter(parameter) {
+    const results = this.setParameters([parameter]);
+    return results[0];
+  }
+
+  /**
+   * Replace a list of declared parameters.
+   * 
+   * Declared parameters are replaced in the order they are provided and
+   * a ParameterEvent is published for each individual parameter change.
+   * 
+   * If an error occurs, the process is stopped and returned. Parameters
+   * set before an error remain unchanged.
+   * 
+   * @param {Parameter[]} parameters - The parameters to set.
+   * @return {rclnodejs.rcl_interfaces.msg.SetParameterResult[]} - A list of SetParameterResult, one for each parameter that was set.
+   */
+  setParameters(parameters = []) {
+    return parameters.map(parameter =>
+      this.setParametersAtomically([parameter])
+    );
+  };
+
+  /**
+   * Repalce a list of declared parameters atomically.
+   * 
+   * Declared parameters are replaced in the order they are provided.
+   * A single ParameterEvent is published collectively for all changed
+   * parameters.
+   * 
+   * If an error occurs, the process stops immediately. All parameters updated to 
+   * the point of the error are reverted to their previous state.
+   * 
+   * @param {Parameter[]} parameters - The parameters to set.
+   * @return {rclnodejs.rcl_interfaces.msg.SetParameterResult} - describes the result of setting 1 or more parameters.
+   */
+  setParametersAtomically(parameters = []) {
+    return this._setParametersAtomically(parameters);
+  };
+
+  /**
+   * Internal method for updating parameters atomically.
+   * 
+   * @param {Paramerter[]} parameters - The parameters to update.
+   * @param {boolean} declareParameterMode - When true parameters are being declared; 
+   *    otherwise they are being changed.
+   * @return {SetParameterResult} - A single collective result.
+   */
+  _setParametersAtomically(parameters = [], declareParameterMode = false) {
+    let result = this._validateParameters(parameters, declareParameterMode);
+    if (!result.successful) {
+      return result;
+    }
+
+    // collectively track updates to parameters for use 
+    // when publishing a ParameterEvent
+    const newParameters = [];
+    const changedParameters = [];
+    const deletedParameters = [];
+
+    for (const parameter of parameters) {
+      if (parameter.type == ParameterType.PARAMETER_NOT_SET) {
+        this.undeclareParameter(parameter.name);
+        deletedParameters.push(parameter);
+      } else {
+        this._parameters.set(parameter.name, parameter);
+        if (declareParameterMode) {
+          newParameters.push(parameter);
+        } else {
+          changedParameters.push(parameter);
+        }
+      }
+    }
+
+    // create ParameterEvent
+    const parameterEvent = new(loader.loadInterface(PARAMETER_EVENT_MSG_TYPE));
```

Comment:
Unless I'm missing something, since loadInterface returns a class type shouldn't this be `new loader.loadInterface(PARAMETER_EVENT_MSG_TYPE)()`. Do you not need the () in ES6?

Reply:
Good question. Originally I had the snippet as you shared which results in this runtime error: `TypeError: loader.loadInterface is not a constructor`.  After seeing this error msg I added the outer parens to force it to work.